File: ModForge/bot/cogs/welcome.py
# -*- coding: utf-8 -*-
"""ModForge Welcome Cog – Welcome/Leave System, Auto-Roles, Sticky Roles."""
import discord
from discord.ext import commands
from discord import app_commands

from bot.config import log
from bot.embed_config import get_embed


class WelcomeCog(commands.Cog):
    """Welcome/Leave Nachrichten, Auto-Roles, Sticky Roles."""

    def __init__(self, bot):
        self.bot = bot

    async def send_welcome(self, member: discord.Member):
        """Sendet die Welcome-Nachricht (aufgerufen vom Events-Cog)."""
        cfg = self.bot.db.get_config(member.guild.id)
        welcome = cfg.get("welcome", {})
        if not welcome.get("enabled") or not welcome.get("channel_id"):
            return

        channel = member.guild.get_channel(welcome["channel_id"])
        if not channel:
            return

        # Zentrales Embed verwenden
        embed = get_embed("welcome", guild=member.guild, user=member, bot=self.bot)
        
        # Custom Override aus Config
        if welcome.get("embed_title"):
            embed.title = welcome["embed_title"].replace("{mention}", member.mention).replace("{user}", str(member))
        if welcome.get("embed_description"):
            embed.description = welcome["embed_description"].replace("{mention}", member.mention).replace("{user}", str(member))

        try:
            content = member.mention if welcome.get("mention") else None
            await channel.send(content=content, embed=embed)
        except (discord.Forbidden, discord.HTTPException):
            pass

        # DM
        if welcome.get("dm_enabled") and welcome.get("dm_description"):
            dm_desc = welcome["dm_description"].replace("{mention}", member.mention).replace("{user}", str(member))
            try:
                from bot.bot import safe_dm
                dm_embed = get_embed("welcome_dm", guild=member.guild, user=member, bot=self.bot)
                await safe_dm(member, dm_embed)
            except Exception:
                pass

        # Auto-Roles
        add_roles = welcome.get("add_roles", [])
        for role_id in add_roles:
            role = member.guild.get_role(role_id)
            if role:
                try:
                    await member.add_roles(role, reason="Auto-Role bei Join")
                except (discord.Forbidden, discord.HTTPException):
                    pass

    async def send_leave(self, member: discord.Member):
        """Sendet die Leave-Nachricht."""
        cfg = self.bot.db.get_config(member.guild.id)
        leave = cfg.get("leave", {})
        if not leave.get("enabled") or not leave.get("channel_id"):
            return

        channel = member.guild.get_channel(leave["channel_id"])
        if not channel:
            return

        embed = get_embed("leave", guild=member.guild, user=member, bot=self.bot)
        
        if leave.get("embed_title"):
            embed.title = leave["embed_title"].replace("{user}", str(member))
        if leave.get("embed_description"):
            embed.description = leave["embed_description"].replace("{user}", str(member))

        try:
            await channel.send(embed=embed)
        except (discord.Forbidden, discord.HTTPException):
            pass

    async def restore_sticky_roles(self, member: discord.Member):
        """Stellt Sticky-Roles wieder her."""
        cfg = self.bot.db.get_config(member.guild.id)
        sticky = cfg.get("sticky_roles", [])
        if not sticky:
            return
        try:
            doc = await self.bot.db.data.find_one(
                {"guild_id": member.guild.id, "user_id": member.id, "type": "sticky_roles"})
            if doc and doc.get("roles"):
                for role_id in doc["roles"]:
                    if role_id in sticky:
                        role = member.guild.get_role(role_id)
                        if role:
                            try:
                                await member.add_roles(role, reason="Sticky-Role wiederhergestellt")
                            except (discord.Forbidden, discord.HTTPException):
                                pass
        except Exception as e:
            log.debug(f"Sticky-Role restore error: {e}")

    # ── COMMANDS ──
    @app_commands.command(name="welcome_channel", description="Setzt den Welcome-Kanal")
    @app_commands.describe(channel="Der Kanal")
    @app_commands.default_permissions(administrator=True)
    async def slash_welcome_channel(self, interaction: discord.Interaction, channel: discord.TextChannel):
        cfg = self.bot.db.get_config(interaction.guild.id)
        welcome = cfg.setdefault("welcome", {})
        welcome["channel_id"] = channel.id
        welcome["enabled"] = True
        await self.bot.db.aset_config(interaction.guild.id, cfg)
        await interaction.response.send_message(f"✅ Welcome-Kanal auf {channel.mention} gesetzt.", ephemeral=True)

File: modforge/bot/cogs/welcome.js
const { SlashCommandBuilder, PermissionFlagsBits } = require('discord.js');
const { getEmbed } = require('../embed_config');

function replaceVars(text, member) {
  return String(text || '')
    .replaceAll('{mention}', `${member}`)
    .replaceAll('{user}', String(member.user?.tag || member))
    .replaceAll('{username}', String(member.user?.username || member.displayName || member))
    .replaceAll('{server}', member.guild?.name || '')
    .replaceAll('{count}', String(member.guild?.memberCount || 0));
}

function styleWelcomeEmbed(embed, config, member) {
  if (config.embed_title) embed.setTitle(replaceVars(config.embed_title, member));
  if (config.embed_description) embed.setDescription(replaceVars(config.embed_description, member));
  if (config.embed_color) embed.setColor(config.embed_color);
  if (config.embed_image) embed.setImage(config.embed_image);
  if (config.embed_thumbnail && member.user?.displayAvatarURL) embed.setThumbnail(member.user.displayAvatarURL({ size: 256 }));
  return embed;
}

async function sendWelcome(bot, member) {
  const cfg = await bot.db.fetchConfig(member.guild.id);
  const welcome = cfg.welcome || {};
  for (const roleId of welcome.add_roles || []) await member.roles.add(String(roleId), 'Auto-Role bei Join').catch(() => null);
  if (welcome.dm_enabled && welcome.dm_description) {
    const dmEmbed = styleWelcomeEmbed(getEmbed('welcome', { guild: member.guild, user: member, bot }), { ...welcome, embed_description: welcome.dm_description }, member);
    await member.send({ embeds: [dmEmbed] }).catch(() => null);
  }
  if (!welcome.enabled || !welcome.channel_id) return;
  const channel = await member.guild.channels.fetch(String(welcome.channel_id)).catch(() => null);
  if (!channel?.isTextBased?.()) return;
  const embed = styleWelcomeEmbed(getEmbed('welcome', { guild: member.guild, user: member, bot }), welcome, member);
  await channel.send({ content: welcome.mention ? `${member}` : null, embeds: [embed] }).catch(() => null);
}

async function sendLeave(bot, member) {
  const cfg = await bot.db.fetchConfig(member.guild.id);
  const leave = cfg.leave || {};
  if (!leave.enabled || !leave.channel_id) return;
  const channel = await member.guild.channels.fetch(leave.channel_id).catch(() => null);
  if (!channel?.isTextBased?.()) return;
  const embed = styleWelcomeEmbed(getEmbed('leave', { guild: member.guild, user: member, bot }), leave, member);
  await channel.send({ embeds: [embed] }).catch(() => null);
}

async function restoreStickyRoles(bot, member) {
  const cfg = await bot.db.fetchConfig(member.guild.id);
  const sticky = (cfg.sticky_roles || []).map(String);
  if (!sticky.length) return;
  const doc = await bot.db.data.findOne({ guild_id_str: String(member.guild.id), user_id: String(member.id), type: 'sticky_roles' }).catch(() => null)
    || await bot.db.data.findOne({ guild_id: Number(member.guild.id), user_id: String(member.id), type: 'sticky_roles' }).catch(() => null);
  if (doc?.roles) for (const roleId of doc.roles.map(String)) if (sticky.includes(roleId)) await member.roles.add(roleId, 'Sticky-Role wiederhergestellt').catch(() => null);
}

async function saveStickyRoles(bot, member) {
  const cfg = await bot.db.fetchConfig(member.guild.id);
  const sticky = new Set((cfg.sticky_roles || []).map(String));
  if (!sticky.size) return;
  const roles = member.roles.cache.map(role => String(role.id)).filter(id => sticky.has(id));
  await bot.db.data.updateOne(
    { type: 'sticky_roles', guild_id_str: String(member.guild.id), user_id: String(member.id) },
    { $set: { type: 'sticky_roles', guild_id_str: String(member.guild.id), user_id: String(member.id), roles, updated_at: new Date() } },
    { upsert: true },
  ).catch(() => null);
}

async function applyAutoNickname(bot, member) {
  if (member.user.bot || member.manageable === false) return;
  const cfg = await bot.db.fetchConfig(member.guild.id);
  const settings = cfg.auto_nickname || {};
  if (!settings.enabled) return;
  const exempt = new Set((settings.exempt_roles || []).map(item => String(item?.id || item)));
  if (member.roles.cache.some(role => exempt.has(String(role.id)))) return;
  const rule = (settings.rules || []).find(item => member.roles.cache.has(String(item.role_id)));
  if (!rule) return;
  const base = member.user.globalName || member.user.username || member.displayName;
  const nickname = `${rule.prefix || ''}${base}${rule.suffix || ''}`.slice(0, 32);
  await member.setNickname(nickname || null, 'ModForge Auto-Nickname').catch(() => null);
}

const commands = [{
  data: new SlashCommandBuilder().setName('welcome_channel').setDescription('Setzt den Welcome-Kanal')
    .setDefaultMemberPermissions(PermissionFlagsBits.Administrator)
    .addChannelOption(o => o.setName('channel').setDescription('Der Kanal').setRequired(true)),
  async execute(bot, interaction) {
    const channel = interaction.options.getChannel('channel', true);
    const cfg = await bot.db.fetchConfig(interaction.guild.id);
    cfg.welcome = cfg.welcome || {};
    cfg.welcome.channel_id = channel.id;
    cfg.welcome.enabled = true;
    await bot.db.setConfig(interaction.guild.id, cfg);
    await interaction.reply({ content: `✅ Welcome-Kanal auf ${channel} gesetzt.`, ephemeral: true });
  }
}];

const events = [
  { name: 'guildMemberAdd', async execute(bot, member) { await restoreStickyRoles(bot, member); await sendWelcome(bot, member); await applyAutoNickname(bot, member); } },
  { name: 'guildMemberRemove', async execute(bot, member) { await saveStickyRoles(bot, member); await sendLeave(bot, member); } },
];

module.exports = { commands, events, sendWelcome, sendLeave, restoreStickyRoles, saveStickyRoles, applyAutoNickname };
